src/components/user.js

```diff
@@ -15,69 +26,120 @@ export default class User extends React.Component {
-        this.handleClickCompany = this.handleClickCompany.bind(this);
-        this.handleChange       = this.handleChange.bind(this);
-        this.handleClickEdit    = this.handleClickEdit.bind(this);
-        this.handleClickSubmit  = this.handleClickSubmit.bind(this);
+    _validateFields (user) {
+        return {
+            [FIELD_NAMES.name]:  this._validateField(FIELD_NAMES.name, user.name),
+            [FIELD_NAMES.phone]: this._validateField(FIELD_NAMES.phone, user.phone),
+            [FIELD_NAMES.email]: this._validateField(FIELD_NAMES.email, user.email),
+        };
     }
 
-    handleClickAddress (e) {
+    _handleClickAddress = e => {
         this.setState({ showAddress: !this.state.showAddress });
         e.preventDefault();
-    }
+    };
 
-    handleClickCompany (e) {
+    _handleClickCompany = e => {
         this.setState({ showCompany: !this.state.showCompany });
         e.preventDefault();
-    }
+    };
 
-    handleClickEdit (e) {
+    _handleClickEdit = e => {
         const { currentUser, wantEdit } = this.state;
 
         if (wantEdit) {
             const newUser = new UserInfo(currentUser.info, currentUser.address, currentUser.company);
 
             UserAction.updateUser(newUser);
         }
-
         this.setState({ wantEdit: !wantEdit });
         e.preventDefault();
-    }
+    };
 
-    handleChange (e, type) {
-        const { currentUser, wantEdit } = this.state;
-        const {isNewUser} = this.props;
+    _handleChange = (e, obj) => {
+        let { currentUser, wantEdit, formErrors, showUserError, addDuplicate } = this.state;
+        let { isNewUser }                                        = this.props;
 
         if (wantEdit || isNewUser) {
-            currentUser[type][e.target.name] = e.target.value;
-            this.setState({ currentUser });
+            const name  = e.target.name;
+            const value = e.target.value;
+
+            currentUser[obj][name] = value;
+            formErrors[name]       = this._validateField(name, value);
+            if (name === FIELD_NAMES.name) {
+                showUserError = false;
+                addDuplicate = false;
+            }
+            this.setState({ currentUser, formErrors, showUserError, addDuplicate });
         }
         e.preventDefault();
-    }
-
-    handleClickSubmit () {
-        const { currentUser } = this.state;
-        const newUser         = new UserInfo(currentUser.info, currentUser.address, currentUser.company);
-
-        UserAction.addNewUser(newUser);
-    }
+    };
+
+    _handleClickSubmit = e => {
+        const { currentUser, addDuplicate, formErrors } = this.state;
+        const newUser                                   = new UserInfo(currentUser.info, currentUser.address, currentUser.company);
+        const resultValid                               = this._validateFields(currentUser.info);
+        const isUserInfoValid                           = !resultValid[FIELD_NAMES.name] &&
```

Comment:
see we hava a very complex check that user info is valid. We can create a method or getter `isUserInfoValid (validInfo)` with ab oolean return value and reuse it for `resultValid` here and for `formErrors` in the `render` method